dotnet/src/Microsoft.Agents.AI.A2A/A2AAgent.cs

```diff
@@ -194,5 +227,93 @@ private static void UpdateThreadConversationId(A2AAgentThread? thread, string? c
 
         // Assign a server-generated context Id to the thread if it's not already set.
         thread.ContextId ??= contextId;
+        thread.TaskId = taskId;
+    }
+
+    private static AgentMessage CreateA2AMessage(A2AAgentThread typedThread, IEnumerable<ChatMessage> messages)
+    {
+        var a2aMessage = messages.ToA2AMessage();
+
+        // Linking the message to the existing conversation, if any.
+        // See: https://github.com/a2aproject/A2A/blob/main/docs/topics/life-of-a-task.md#group-related-interactions
+        a2aMessage.ContextId = typedThread.ContextId;
+
+        // Link the message as a follow-up to an existing task, if any.
+        // See: https://github.com/a2aproject/A2A/blob/main/docs/topics/life-of-a-task.md#task-refinements
+        a2aMessage.ReferenceTaskIds = typedThread.TaskId is null ? null : [typedThread.TaskId];
+
+        return a2aMessage;
+    }
+
+    private static A2AContinuationToken? GetContinuationToken(IEnumerable<ChatMessage> messages, AgentRunOptions? options = null)
+    {
+        if (options?.ContinuationToken is ResponseContinuationToken token)
+        {
+            if (messages.Any())
+            {
+                throw new InvalidOperationException("Messages are not allowed when continuing a background response using a continuation token.");
+            }
+
+            return A2AContinuationToken.FromToken(token);
+        }
+
+        return null;
+    }
+
+    private static A2AContinuationToken? CreateContinuationToken(string taskId, TaskState state)
+    {
+        if (state == TaskState.Submitted || state == TaskState.Working)
+        {
+            return new A2AContinuationToken(taskId);
+        }
+
+        return null;
+    }
+
+    private AgentRunResponseUpdate ConvertToAgentResponseUpdate(AgentMessage message)
+    {
+        return new AgentRunResponseUpdate
+        {
+            AgentId = this.Id,
+            ResponseId = message.MessageId,
+            RawRepresentation = message,
+            Role = ChatRole.Assistant,
+            MessageId = message.MessageId,
+            Contents = message.Parts.ConvertAll(part => part.ToAIContent()),
+            AdditionalProperties = message.Metadata.ToAdditionalProperties(),
+        };
+    }
+
+    private AgentRunResponseUpdate ConvertToAgentResponseUpdate(AgentTask task)
+    {
+        return new AgentRunResponseUpdate
+        {
+            AgentId = this.Id,
+            ResponseId = task.Id,
+            RawRepresentation = task,
+            Role = ChatRole.Assistant,
+            Contents = task.ToAIContents(),
+            AdditionalProperties = task.Metadata.ToAdditionalProperties(),
```

Comment:
I think `Metadata` is nullable.

```suggestion
            AdditionalProperties = task.Metadata?.ToAdditionalProperties(),
```